src/ynEIGEN/AssetRegistry.sol

```diff
@@ -0,0 +1,342 @@
+        revert AssetNotFound(address(asset));
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ASSET VALUE  -------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /**
+     * @notice This function calculates the total assets of the contract
+     * @dev It iterates over all the assets in the contract, gets the latest price for each asset from the oracle, 
+     * multiplies it with the balance of the asset and adds it to the total
+     * @return total The total assets of the contract in the form of uint
+     */
+    function totalAssets() public view returns (uint256) {
+        uint256 total = 0;
+
+        uint256[] memory depositedBalances = getAllAssetBalances();
+
+        for (uint256 i = 0; i < assets.length; i++) {
+            uint256 balanceInUnitOfAccount = convertToUnitOfAccount(assets[i], depositedBalances[i]);
+            total += balanceInUnitOfAccount;
+        }
+        return total;
+    }
+
+    /**
+     * @notice Retrieves the total balances of all assets managed by the contract, both held directly and managed through strategies.
+     * @dev This function aggregates the balances of each asset held directly by the contract and in each LSDStakingNode, 
+     * including those managed by strategies associated with each asset.
+     * @return assetBalances An array of the total balances for each asset, indexed in the same order as the `assets` array.
+     */
+    function getAllAssetBalances()
+        public
+        view
+        returns (uint256[] memory assetBalances)
+    {
+        uint256 assetsCount = assets.length;
+
+        assetBalances = new uint256[](assetsCount);
+
+        // populate with the ynEigen balances
+        assetBalances = ynEigen.assetBalances(assets);
+
+        uint256[] memory stakedAssetBalances = eigenStrategyManager.getStakedAssetsBalances(assets);
+
+        if (stakedAssetBalances.length != assetsCount) {
+            revert LengthMismatch(assetsCount, stakedAssetBalances.length);
+        }
+
+        for (uint256 i = 0; i < assetsCount; i++) {
+            assetBalances[i] += stakedAssetBalances[i];
+        }
+    }
+
+    /**
+     * @notice Converts the amount of a given asset to its equivalent value in the unit of account of the vault.
+     * @dev This function takes into account the decimal places of the asset to ensure accurate conversion.
+     * @param asset The ERC20 token to be converted to the unit of account.
+     * @param amount The amount of the asset to be converted.
+     * @return uint256 equivalent amount of the asset in the unit of account.
+     */
+    function convertToUnitOfAccount(IERC20 asset, uint256 amount) public view returns (uint256) {
+        uint256 assetRate = rateProvider.rate(address(asset));
+        uint8 assetDecimals = IERC20Metadata(address(asset)).decimals();
```

Comment:
this really really shouldn't happen but we could add some defense-in-depth here by adding: `assert(assetDecimals <= 18)`

Reply:
I don't think we should, the current code works for any number of decimals